### PR TITLE
checker: reduces the probability of deleting normal peers when the store becomes unavailable (#7249)

### PR DESCRIPTION
**Pull request by lhy1024**
It is a pick for https://github.com/tikv/pd/pull/7240, https://github.com/tikv/pd/pull/7294 and https://github.com/tikv/pd/pull/7315

### What problem does this PR solve?

Issue Number: Close #7249

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by nolouch — APPROVED** [review]

File: server/schedule/checker/rule_checker_test.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ package checker
 import (
 	"context"
 	"fmt"
+
```
Comment (rleungx) [inline server/schedule/checker/rule_checker_test.go:20]: empty line

File: server/schedule/checker/rule_checker.go [hunk 2]
```diff
@@ -404,24 +404,45 @@ func (c *RuleChecker) fixOrphanPeers(region *core.RegionInfo, fit *placement.Reg
...
+			return true
+		}
+		return store.IsDisconnected()
+	}
+
+	checkDownPeer := func(peers []*metapb.Peer) (*metapb.Peer, bool) {
+		for _, p := range peers {
+			if isUnhealthyPeer(p.GetId()) {
+				// make sure is down peer.
+				if region.GetDownPeer(p.GetId()) != nil {
```
Comment (rleungx) [inline server/schedule/checker/rule_checker.go:422]: Is it duplicated?

Comment (lhy1024) [reply server/schedule/checker/rule_checker.go:422]: `isUnhealthyPeer` contains `isDownPeer` and `isPendingPeer`

**Review by bufferflies — APPROVED** [review]

File: server/schedule/checker/rule_checker.go (hunk 2 above)
Comment (lhy1024) [reply server/schedule/checker/rule_checker.go:422]: it will do in https://github.com/tikv/pd/pull/7342

**Pull request merged**